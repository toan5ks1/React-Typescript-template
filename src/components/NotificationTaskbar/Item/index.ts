export * from "./MaterialItem";
export * from "./MasterImportItem";
