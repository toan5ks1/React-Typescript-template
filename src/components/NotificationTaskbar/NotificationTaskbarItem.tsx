import { Box } from "@mui/material";

export interface NotificationTaskbarItemProps {
    onClick: () => void;
    content: string;
}

export const NotificationTaskbarItem = (props: NotificationTaskbarItemProps) => {
    const { onClick, content } = props;
    return (
        <Box
            sx={(theme) => ({
                width: theme.spacing(28),
                height: theme.spacing(8.375),
                padding: theme.spacing(3, 2),
                display: "flex",
                alignItems: "center",
                borderRadius: "8px 8px 0px 0px",
                border: "1px solid #E6E8EF",
                cursor: "pointer",
                "&:not(:first-of-type)": {
                    marginLeft: theme.spacing(0.5),
                },
                backgroundColor: theme.palette.background.paper,
            })}
            onClick={onClick}
            data-testid="NotificationTaskbarItem__root"
        >
            <Box
                whiteSpace="nowrap"
                overflow="hidden"
                textOverflow="ellipsis"
                data-testid="NotificationTaskbarItem__name"
                data-content={content}
            >
                {content}
            </Box>
        </Box>
    );
};

export default NotificationTaskbarItem;
