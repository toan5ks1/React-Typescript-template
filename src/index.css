body {
    margin: 0;
    font-family: "Roboto", sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    overflow-x: hidden;
}

code {
    font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace;
}

body,
html,
#root {
    margin: 0;
    padding: 0;
    width: 100%;
    min-height: 100vh;
    background-color: #fff;
}

#root > div:first-child {
    margin: 0;
    padding: 0;
    width: 100%;
    min-height: 100vh;
    box-sizing: border-box;
}

.list-page > div > div:last-child {
    overflow: auto;
}

table td {
    max-width: 300px;
    overflow-wrap: break-word;
}

/* Override css background input autocomplete - Designer wants it */
input:-webkit-autofill,
input:-webkit-autofill:hover,
input:-webkit-autofill:focus,
textarea:-webkit-autofill,
textarea:-webkit-autofill:hover,
textarea:-webkit-autofill:focus,
select:-webkit-autofill,
select:-webkit-autofill:hover,
select:-webkit-autofill:focus {
    -webkit-box-shadow: 0 0 0px 1000px #fff inset;
    transition: background-color 5000s ease-in-out 0s;
}

/* styled not affectted, so we use this, check src/components/BrightcoveVideo/BrightcoveVideo.tsx */
.bc-player-default_default {
    width: 100%;
    height: 100%;
}
