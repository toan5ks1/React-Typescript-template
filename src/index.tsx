import "src/styles/mui-classname-setup";

import { StrictMode } from "react";

import { enableMapSet } from "immer";
import ReactDOM from "react-dom";
import enableSidebarMethod from "src/squads/architecture/internals/sidebar";
import enableManaEventEmitter from "src/squads/root/src/internals/mana-event-emitter";

import App from "./app";
import AppProvider from "./app/AppProvider";
import "./index.css";
import { enableReport } from "./internals/stackdriver";
import { enableLogger } from "./internals/warner";

enableMapSet();
enableReport();
enableLogger();

enableSidebarMethod();
enableManaEventEmitter();

ReactDOM.render(
    <StrictMode>
        <AppProvider>
            <App />
        </AppProvider>
    </StrictMode>,
    document.getElementById("root")
);
