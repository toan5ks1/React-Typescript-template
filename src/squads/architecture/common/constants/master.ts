// Cloned from src/squads/payment/constants/master.ts
// TODO: Consider remove this file and move everything to Payment later

// Payment
export enum OrderManagementCategory {
    Discount = "discount",
    Fee = "fee",
    LeavingReason = "leavingReason",
    Material = "material",
    Package = "package",
    PackageCourseMaterial = "packageCourseMaterial",
    PackageCourseFee = "packageCourseFee",
    PackageQuantityTypeMapping = "packageQuantityTypeMapping",
    ProductAccountingCategory = "productAccountingCategory",
    ProductCourse = "productCourse",
    ProductGrade = "productGrade",
    ProductLocation = "productLocation",
    ProductPrice = "productPrice",
    Tax = "tax",
    AccountingCategory = "accountingCategory",
    BillingRatio = "billingRatio",
    BillingSchedule = "billingSchedule",
    BillingSchedulePeriod = "billingSchedulePeriod",
}

// User
enum UserSchoolGroupCategory {
    SchoolLevel = "schoolLevel",
    School = "school",
    SchoolCourse = "schoolCourse",
    SchoolLevelGrade = "schoolLevelGrade",
}

enum UserTagGroupCategory {
    StudentTag = "studentTag",
    ParentTag = "parentTag",
}

enum UserBankGroupCategory {
    Bank = "bank",
    BankBranch = "bankBranch",
}

enum UserTimesheetGroupCategory {
    TimesheetConfig = "timesheetConfig",
}

// Architecture
export enum ArchitectureReserveCategory {
    Grade = "grades",
}
export enum ArchitectureCategory {
    Class = "class",
    Location = "location",
    LocationType = "locationType",
}

export type MasterCategoryType =
    | OrderManagementCategory
    | UserSchoolGroupCategory
    | UserTagGroupCategory
    | UserBankGroupCategory
    | UserTimesheetGroupCategory
    | ArchitectureReserveCategory
    | ArchitectureCategory;
