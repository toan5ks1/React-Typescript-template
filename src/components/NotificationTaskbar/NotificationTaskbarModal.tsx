import { Entities } from "src/common/constants/enum";
import { isMaterialConverting } from "src/store/lesson-convert";
import { FileImportMaster } from "src/store/master/types";

import { Box, LinearProgress } from "@mui/material";
import BaseDialog, {
    BaseDialogProps,
} from "src/squads/syllabus/components/LegacyDialogs/BaseDialog";

import { arrayHasItem } from "../../common/utils/other";
import { Material } from "../../store/lesson-convert/lesson-convert-types";
import { MasterImportItem, MaterialItem } from "./Item";

type ItemType = Material | FileImportMaster;
export interface NotificationTaskbarModalProps extends BaseDialogProps {
    items?: ItemType[];
}

function isMaterial(item: ItemType): item is Material {
    return "name" in item;
}

function isImportedMasterFile(item: ItemType): item is FileImportMaster {
    const typeTaskbar = item.id.split("_")?.[0];

    return typeTaskbar === Entities.MASTERS;
}

function shouldRenderNotificationItem(item: ItemType) {
    if (isImportedMasterFile(item)) return true;

    if (!isMaterial(item)) return false;
    if (!("status" in item)) return false;

    if (isMaterial(item) && !isMaterialConverting(item.status)) return false;

    return true;
}

export const NotificationTaskbarModal = ({
    items = [],
    ...rest
}: NotificationTaskbarModalProps) => {
    if (!arrayHasItem(items)) return null;

    return (
        <BaseDialog {...rest}>
            <div data-testid="NotificationTaskbarModal__list">
                {items.map((item: ItemType, index) => {
                    if (!shouldRenderNotificationItem(item)) return null;

                    return (
                        <Box sx={(theme) => ({ marginBottom: theme.spacing(3) })} key={index}>
                            <Box
                                sx={(theme) => ({
                                    display: "flex",
                                    alignItems: "center",
                                    marginBottom: theme.spacing(1),
                                })}
                            >
                                {isMaterial(item) && <MaterialItem {...item} />}
                                {isImportedMasterFile(item) && (
                                    <MasterImportItem fileName={item.file.name} />
                                )}
                            </Box>

                            <LinearProgress data-testid="NotificationTaskbarModal__progress" />
                        </Box>
                    );
                })}
            </div>
        </BaseDialog>
    );
};

export default NotificationTaskbarModal;
