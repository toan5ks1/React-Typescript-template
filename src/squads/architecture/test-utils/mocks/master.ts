import { MIMETypes } from "src/common/constants/enum";
import {
    OrderManagementCategory,
    ArchitectureCategory,
} from "src/squads/architecture/common/constants/master";
import { NsArchitecture_Mastermgmt_LocationsService } from "src/squads/architecture/service/mastermgmt/locations-service/types";

import { generateFiles } from "./file";

export const createMockCSVFiles = (count: number = 1) => {
    const fileSchemas = Array.from(Array(count), (_, i) => i + 1).map((number) => ({
        size: 300 * number,
        fileProperty: {
            type: MIMETypes.CSV,
        },
    }));

    return generateFiles(fileSchemas);
};

export const createMockLocationImportPayload =
    (): NsArchitecture_Mastermgmt_LocationsService.ImportMasterDataRequest["payload"] => {
        const file = createMockCSVFiles()[0];

        return {
            file,
            category: ArchitectureCategory.Location,
        };
    };

export const createMockLocationImportPayloadInvalid =
    (): NsArchitecture_Mastermgmt_LocationsService.ImportMasterDataRequest["payload"] => {
        const file = createMockCSVFiles()[0];

        return {
            file,
            category: OrderManagementCategory.AccountingCategory,
        };
    };
