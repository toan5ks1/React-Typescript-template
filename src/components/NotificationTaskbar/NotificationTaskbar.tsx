import { useCallback, useEffect, useMemo, useState } from "react";

import { useSelector } from "react-redux";
import { Entities } from "src/common/constants/enum";
import { getConvertMaterialLessonSelector } from "src/store/lesson-convert";
import { importedFilesSelector } from "src/store/master/selectors";

import { Box } from "@mui/material";

import { Material } from "../../store/lesson-convert/lesson-convert-types";
import NotificationTaskbarItem from "./NotificationTaskbarItem";
import NotificationTaskbarModal from "./NotificationTaskbarModal";

import useDialog from "src/hooks/useDialog";
import useResourceTranslate from "src/hooks/useResourceTranslate";
import useTranslate from "src/hooks/useTranslate";

const NotificationTaskbar = () => {
    const t = useTranslate();
    const tMaster = useResourceTranslate(Entities.MASTERS);

    const { open, onClose, onOpen } = useDialog();
    const [key, setKey] = useState<string>();
    const [title, setTitle] = useState<string>(t("resources.lesson_syllabus.uploadMaterials"));
    const materialStores = useSelector(getConvertMaterialLessonSelector);

    const masterImportFiles = useSelector(importedFilesSelector);

    const materialsUploading: Material[] = useMemo(() => {
        let result: Material[] = [];
        materialStores
            .filter((e) => e.courseId === key)
            .map((e) => {
                if (e.material && e.material.length) {
                    result = [...result, ...e.material];
                }
            });
        return result;
    }, [key, materialStores]);

    const mastersUploading = useMemo(
        () => masterImportFiles.filter((file) => file.id === key),
        [key, masterImportFiles]
    );

    const modalItems = useMemo(
        () => [...materialsUploading, ...mastersUploading],
        [materialsUploading, mastersUploading]
    );

    const handleOnOpen = useCallback(
        (key: string, title?: string) => {
            setKey(key);
            title && setTitle(title);
            onOpen();
        },
        [onOpen]
    );

    const handleOnClose = useCallback(() => {
        onClose();
        setKey("");
    }, [onClose]);

    useEffect(() => {
        if (!open || modalItems.length) return;
        handleOnClose();
    }, [modalItems.length, open, handleOnClose]);

    return (
        <>
            <NotificationTaskbarModal
                open={open}
                onClose={handleOnClose}
                title={title}
                minWidth="770px"
                maxWidth={false}
                items={modalItems}
            />
            <Box
                sx={(theme) => ({
                    position: "fixed",
                    bottom: "0",
                    display: "inline-block",
                    height: "64px",
                    zIndex: 1201,
                    paddingLeft: theme.spacing(4),
                })}
            >
                <Box display="flex" data-testid="NotificationTaskbar__list">
                    {materialStores.map((lessonGroup) => {
                        const { lessonGroupId, courseId } = lessonGroup;

                        const content = t("resources.courses.lessonConvert.converting", {
                            fileName: lessonGroupId,
                        });
                        return (
                            <NotificationTaskbarItem
                                key={`${courseId}.${lessonGroupId}`}
                                content={content}
                                onClick={() =>
                                    handleOnOpen(
                                        courseId,
                                        t("resources.lesson_syllabus.uploadMaterials")
                                    )
                                }
                            />
                        );
                    })}
                    {masterImportFiles.map(({ id, file }) => (
                        <NotificationTaskbarItem
                            key={id}
                            content={tMaster("status.fileUploading", {
                                fileName: file.name,
                            })}
                            onClick={() => {
                                handleOnOpen(id);
                            }}
                        />
                    ))}
                </Box>
            </Box>
        </>
    );
};

export default NotificationTaskbar;
