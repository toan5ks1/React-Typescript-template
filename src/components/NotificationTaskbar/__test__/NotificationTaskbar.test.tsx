import { createMockMasterImportFile } from "src/squads/payment/test-utils/mocks/master";
import { TestApp, TestThemeProvider } from "src/test-utils";

import NotificationTaskbar from "../NotificationTaskbar";

import { render, cleanup, RenderResult, screen } from "@testing-library/react";

const mockMasterFile = createMockMasterImportFile("", true);

afterAll(() => {
    jest.resetAllMocks();
    cleanup();
});

jest.mock("src/store/lesson-convert/selectors", () => {
    return {
        getConvertMaterialLessonSelector: () => [
            {
                courseId: "01",
                lessonGroupId: "lesson 01",
                material: [],
            },
            {
                courseId: "01",
                lessonGroupId: "lesson 02",
                material: [],
            },
        ],
    };
});

jest.mock("src/store/master/selectors", () => {
    return {
        importedFilesSelector: () => [mockMasterFile],
    };
});

describe("<NotificationTaskbar />", () => {
    const list = "NotificationTaskbar__list";
    const itemName = "NotificationTaskbarItem__name";

    let wrapper: RenderResult;

    beforeEach(() => {
        wrapper = render(
            <TestApp>
                <TestThemeProvider>
                    <NotificationTaskbar />
                </TestThemeProvider>
            </TestApp>
        );
    });

    it("should match snapshot", () => {
        expect(wrapper.container).toMatchSnapshot();
    });

    it("should match UI", () => {
        expect(screen.getAllByTestId(itemName)[0].textContent).toMatch("lesson 01 converting ...");
        expect(screen.getByTestId(list).children).toHaveLength(3);
    });
});
