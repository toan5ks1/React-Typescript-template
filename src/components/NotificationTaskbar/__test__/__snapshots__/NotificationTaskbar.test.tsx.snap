// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationTaskbar /> should match snapshot 1`] = `
.emotion-0 {
  position: fixed;
  bottom: 0;
  display: inline-block;
  height: 64px;
  z-index: 1201;
  padding-left: 32px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-2 {
  width: 224px;
  height: 67px;
  padding: 24px 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 8px 8px 0px 0px;
  border: 1px solid #E6E8EF;
  cursor: pointer;
  background-color: #fff;
}

.emotion-2:not(:first-of-type) {
  margin-left: 4px;
}

.emotion-3 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
      data-testid="NotificationTaskbar__list"
    >
      <div
        class="MuiBox-root emotion-2"
        data-testid="NotificationTaskbarItem__root"
      >
        <div
          class="MuiBox-root emotion-3"
          data-content="lesson 01 converting ..."
          data-testid="NotificationTaskbarItem__name"
        >
          lesson 01 converting ...
        </div>
      </div>
      <div
        class="MuiBox-root emotion-2"
        data-testid="NotificationTaskbarItem__root"
      >
        <div
          class="MuiBox-root emotion-3"
          data-content="lesson 02 converting ..."
          data-testid="NotificationTaskbarItem__name"
        >
          lesson 02 converting ...
        </div>
      </div>
      <div
        class="MuiBox-root emotion-2"
        data-testid="NotificationTaskbarItem__root"
      >
        <div
          class="MuiBox-root emotion-3"
          data-content="Uploading FILE_NAME ..."
          data-testid="NotificationTaskbarItem__name"
        >
          Uploading FILE_NAME ...
        </div>
      </div>
    </div>
  </div>
</div>
`;
